File: cli/src/mandala_fleet/tui/explorer.py
"""Fleet TUI: members / groups / drift tabs + the action tier.

The VIEWS are strictly read-only — browsing members, groups, and the
deployed-generation drift dashboard never mutates fleet state. Actions
(ping, reboot, deploy) target the cursor's selection and run in pushed
SCREENS, so the mutation boundary is a screen edge the operator crosses
deliberately; reboot and deploy additionally sit behind a confirm modal,
and the playbooks keep their own guards (--limit, k8s drain handling in
the operator's reboot playbook).

Task availability is conventional: ping is ansible ad-hoc; reboot is
offered only when the operator repo ships playbooks/reboot.yaml.
"""

from __future__ import annotations

import os
import subprocess
import time
from collections import deque
from pathlib import Path

from rich.text import Text
from textual.app import App, ComposeResult
from textual.binding import Binding
from textual.containers import Vertical
from textual.widgets import Footer, Header, Static, TabbedContent, TabPane

from .. import drift as drift_mod
from ..inventory import Inventory, surfaces
from ..runner import DeployRun
from .deploy import DeployScreen
from .select_table import SelectTable
from .tasks import ConfirmScreen, RebootScreen, TaskScreen

# Braille spinner frames for the background-job indicator (see the status
# machinery below). One shared frame animates every running job at once.
_SPINNER = "⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏"


def _ansible_dir() -> Path:
    return Path("ansible") if Path("ansible/ansible.cfg").is_file() else Path(".")


class ExplorerApp(App):
    """Fleet explorer: tabs to browse and select, keys to act."""

    TITLE = "mandala — fleet"
    CSS = """
    /* Fill the space between header and footer, and constrain the tables
       to it — an auto-height DataTable grows past the viewport and never
       scrolls. */
    #views { height: 1fr; }
    #members-table, #groups-table, #drift-table { height: 1fr; }
    #drift-hint { dock: bottom; height: 1; padding: 0 1; color: $text-muted; }
    """
    BINDINGS = [
        Binding("q", "quit", "quit"),
        Binding("r", "reload", "reload"),
        Binding("S", "refresh_drift", "refresh drift (survey + eval)"),
        Binding("p", "ping", "ping selection"),
        Binding("R", "reboot", "reboot selection"),
        Binding("D", "deploy", "deploy selection"),
    ]

    def __init__(self, inventory: Inventory) -> None:
        super().__init__()
        self.inventory = inventory
        self.expected: dict[str, str] | None = None
        self._rev: str | None = None
        self._cached_rev: str | None = None
        # Background-job state. `_busy` covers any nix eval in flight (the
        # inventory aggregate on load/reload, or the expected-toplevel eval);
        # `_surveying` the read-only state survey. Both can run at once.
        self._busy = False
        self._surveying = False
        self._survey_n = 0  # snapshots surveyed so far this run
        # Resting bar text shown when no job is running, and the spinner timer.
        self._status = ""
        self._status_sticky = False  # an error holds until the next refresh
        self._spin = 0
        self._spin_timer = None

    # -- status indicators ---------------------------------------------
    #
    # eval and survey run CONCURRENTLY (action_refresh_drift fires both) and
    # the survey usually finishes first. A single status string meant the
    # survey's "drift refreshed" stomped the still-running eval, so you
    # couldn't tell the eval was in flight until the columns moved. Instead
    # each job owns a running flag; while any are set a timer animates one
    # spinner and the bar lists every job still running — the eval indicator
    # stays put after the faster survey lands. With all jobs idle the bar
    # shows the latest resting message (a result, or a sticky error).

    def _set_status(self, msg: str, *, error: bool = False) -> None:
        """Set the resting bar message (shown when nothing is running).

        Errors are sticky: a concurrently-finishing success will not
        overwrite them, so an eval failure survives the survey's "drift
        refreshed". The stickiness clears when the next refresh begins."""
        if error or not self._status_sticky:
            self._status = msg
            self._status_sticky = error
        self._render_status()

    def _render_status(self) -> None:
        jobs: list[str] = []
        if self._busy:
            jobs.append("eval")
        if self._surveying:
            jobs.append(f"survey ({self._survey_n} read)" if self._survey_n else "survey")
        if jobs:
            frame = _SPINNER[self._spin % len(_SPINNER)]
            self.sub_title = "running   " + "   ·   ".join(f"{frame} {j}" for j in jobs)
            self._ensure_spinner()
        else:
            self.sub_title = self._status

    def _ensure_spinner(self) -> None:
        if self._spin_timer is None:
            self._spin_timer = self.set_interval(0.1, self._tick)

    def _tick(self) -> None:
        self._spin += 1
        self._render_status()
        # Stop animating once every job is idle — the resting message needs
        # no ticking, and a live timer would spin forever.
        if not (self._busy or self._surveying) and self._spin_timer is not None:
            self._spin_timer.stop()
            self._spin_timer = None

    # -- layout --------------------------------------------------------

    def compose(self) -> ComposeResult:
        yield Header(show_clock=True)
        with TabbedContent(initial="tab-members", id="views"):
            with TabPane("members", id="tab-members"):
                yield SelectTable(id="members-table", zebra_stripes=True, cursor_type="row")
            with TabPane("groups", id="tab-groups"):
                yield SelectTable(id="groups-table", zebra_stripes=True, cursor_type="row")
            with TabPane("drift", id="tab-drift"):
                yield Vertical(
                    SelectTable(id="drift-table", zebra_stripes=True, cursor_type="row"),
                    Static(id="drift-hint"),
                )
        yield Footer()

    def on_mount(self) -> None:
        members = self.query_one("#members-table", SelectTable)
        members.add_columns("", "member", "platform", "arch", "category", "role", "tags", "ads")
        groups = self.query_one("#groups-table", SelectTable)
        groups.add_columns("", "group", "n", "members")
        drift = self.query_one("#drift-table", SelectTable)
        drift.add_columns("", "member", "status", "current", "expected", "booted", "captured")
        self._load()

    # -- data ----------------------------------------------------------

    def _load(self) -> None:
        """Aggregate eval OFF the UI thread — it takes tens of seconds on
        a real fleet, and blocking on_mount means a gray void instead of
        a first paint."""
        if self._busy:
            return
        self._busy = True
        self._render_status()  # shows the eval spinner
        inv = self.inventory

        def work() -> None:
            try:
                inv.aggregate  # force the cached_property
            except Exception as e:  # noqa: BLE001 — surfaced in the UI
                self.call_from_thread(self._load_failed, str(e))
                return
            # Reuse the rev-keyed expected cache when the contract hasn't
            # moved since the last eval (a mismatch is itself the signal).
            self._rev = drift_mod.repo_rev(inv.flake)
            self._cached_rev, cached = drift_mod.load_expected()
            if drift_mod.cache_fresh(self._cached_rev, self._rev):
                self.expected = cached
            self.call_from_thread(self._fill)

        self.run_worker(work, thread=True, exclusive=True)

    def _load_failed(self, error: str) -> None:
        self._busy = False
        self._set_status(
            f"aggregate eval failed: {error.splitlines()[-1] if error else 'unknown'}",
            error=True,
        )

    def _fill(self) -> None:
        inv = self.inventory

        members = self.query_one("#members-table", SelectTable)
        members.reset_rows()
        for name in sorted(inv.members):
            m = inv.members[name]
            members.add_named_row(
                name,
                name,
                m.get("platform", "?"),
                m.get("architecture", "?"),
                m.get("category", "?"),
                m.get("role") or "-",
                " ".join(m.get("tags", [])),
                surfaces(m),
            )

        groups = self.query_one("#groups-table", SelectTable)
        groups.reset_rows()
        for group, names in sorted(inv.groups.items()):
            groups.add_named_row(group, group, str(len(names)), " ".join(sorted(names)))

        self._fill_drift()
        self._busy = False
        self._set_status(
            f"{len(inv.members)} members, {len(inv.groups)} groups"
            " — space/shift+↑↓ select · p ping · R reboot · D deploy"
        )

    @property
    def _deploy_nodes(self) -> list[str]:
        deploy = self.inventory.aggregate.get("projections", {}).get("deploy", {})
        return list(deploy.get("nodes", []))

    def _fill_drift(self) -> None:
        snapshots = drift_mod.read_snapshots()
        entries = drift_mod.compare(self._deploy_nodes, snapshots, self.expected)
        table = self.query_one("#drift-table", SelectTable)
        table.reset_rows()
        short = lambda p: (p or "").removeprefix("/nix/store/")[:18]
        for e in entries:
            table.add_named_row(
                e.host,
                e.host,
                Text(e.status.value, style=drift_mod.STATUS_STYLE[e.status]),
                short(e.current),
                short(e.expected),
                short(e.booted),
                (e.captured_at or "")[:19],
            )
        hint = "S refresh drift (survey + eval) · R reboot a reboot-pending row"
        if self.expected is not None:
            hint += f"   expected @ {drift_mod.short_rev(self._rev)}"
        elif self._cached_rev is not None:
            hint += (
                f"   contract MOVED since last eval"
                f" (cache @ {drift_mod.short_rev(self._cached_rev)},"
                f" repo @ {drift_mod.short_rev(self._rev)}) — press S"
            )
        else:
            hint += "   (expected NOT evaluated yet — press S)"
        self.query_one("#drift-hint", Static).update(hint)

    # -- selection -----------------------------------------------------

    def _target(self) -> str | None:
        """The action target on the active tab: the multi-selection when
        one exists, else the cursor row. Names are members or groups (==
        ansible groups, one taxonomy one spelling), comma-joined into an
        ansible --limit."""
        active = self.query_one("#views", TabbedContent).active
        table_id = {
            "tab-members": "#members-table",
            "tab-groups": "#groups-table",
            "tab-drift": "#drift-table",
        }.get(active)
        if table_id is None:
            return None
        table = self.query_one(table_id, SelectTable)
        selected = table.selected_names
        if selected:
            return ",".join(selected)
        return table.cursor_name

    # -- actions (pushed screens; views stay read-only) -----------------

    def action_ping(self) -> None:
        target = self._target()
        if target is None:
            return
        self.push_screen(TaskScreen(
            f"ping {target}",
            ["ansible", target, "-m", "ping"],
            _ansible_dir(),
            # Deliberately the DEFAULT stdout callback: --one-line AND the
            # oneline/minimal callbacks are deprecated in core 2.19
            # (removed 2.23) with no core replacement (ansible/ansible
            # #85333, closed not-planned), and community presentation
            # plugins would be a new dependency. The default callback is
            # the only stable surface; the pane wraps and scrolls.
        ))

    def action_reboot(self) -> None:
        target = self._target()
        if target is None:
            return
        # Prefer the operator's wrapper: it carries the controller-side env
        # raw ansible-playbook lacks — the delegated k8s drain pins
        # ANSIBLE_LOCAL_PYTHON_INTERPRETER to a python WITH the kubernetes
        # lib (the global interpreter default does not win for delegated
        # tasks, so without the wrapper the drain fails "kubernetes
        # library is missing").
        import shutil

        if shutil.which("ans-reboot"):
            base = ["ans-reboot", "-l", target]
        elif (_ansible_dir() / "playbooks/reboot.yaml").is_file():
            base = ["ansible-playbook", "playbooks/reboot.yaml", "-l", target]
        else:
            self._set_status("no ans-reboot wrapper or playbooks/reboot.yaml — reboot task unavailable")
            return

        # RebootScreen returns the chosen batch order + drain safety; both
        # ride through the wrapper as extra-vars (ans-reboot/ansible-playbook
        # both forward "$@"). reboot_serial drives the play's `serial`,
        # drain gates the k8s cordon/drain steps.
        def go(opts: dict | None) -> None:
            if not opts:
                return
            argv = base + [
                "-e", f"reboot_serial={opts['serial']}",
                "-e", f"drain={'true' if opts['drain'] else 'false'}",
            ]
            self.push_screen(
                TaskScreen(f"reboot {target}", argv, _ansible_dir()),
                self._after_mutation,
            )

        self.push_screen(RebootScreen(target), go)

    def action_deploy(self) -> None:
        target = self._target()
        if target is None:
            return

        def go(confirmed: bool | None) -> None:
            if confirmed:
                self.push_screen(
                    DeployScreen(DeployRun(limit=target)),
                    self._after_mutation,
                )

        self.push_screen(
            ConfirmScreen(f"Deploy '{target}'?\n(eval-once batch build, then deploy-rs per host with magic rollback)"),
            go,
        )

    # -- maintenance ---------------------------------------------------

    def action_reload(self) -> None:
        self.inventory = Inventory(flake=self.inventory.flake)
        self.expected = None
        self._load()

    def action_refresh_drift(self) -> None:
        """Refresh both drift inputs at once: the expected-toplevel nix
        eval (a background worker) and the read-only state survey (a
        pushed TaskScreen) run CONCURRENTLY, not as two separate
        keystrokes. Each refreshes the drift table as it lands; either
        completion order converges to the same judgement. Also fired
        automatically once a deploy or reboot completes (see
        `_after_mutation`)."""
        self._status_sticky = False  # a fresh refresh clears a stale error
        self.action_eval_expected()  # background worker, returns at once
        self.action_survey()  # pushed TaskScreen, runs alongside the eval

    def _after_mutation(self, rc: int | None) -> None:
        """A deploy/reboot screen just closed. If the run actually ran to
        completion (rc is set — an operator cancel pops with rc None),
        auto-refresh drift so the post-change state is surveyed without
        having to press S. A non-zero rc (a failed/partial run) still
        refreshes — seeing the resulting state is exactly what you want."""
        if rc is not None:
            self.action_refresh_drift()

    def action_eval_expected(self) -> None:
        if self._busy:
            return
        self._busy = True
        self._render_status()  # shows the eval spinner alongside any survey
        nodes = self._deploy_nodes
        flake = self.inventory.flake

        def work() -> None:
            try:
                expected = drift_mod.eval_expected(flake, nodes)
            except Exception as e:  # noqa: BLE001 — surfaced, not raised
                self.call_from_thread(self._drift_done, None, f"eval failed: {e}")
                return
            self._rev = drift_mod.repo_rev(flake)
            self._cached_rev = self._rev
            drift_mod.save_expected(self._rev, expected)
            self.call_from_thread(self._drift_done, expected, None)

        self.run_worker(work, thread=True, exclusive=True)

    def action_survey(self) -> None:
        """Read-only state survey, run in the BACKGROUND rather than as a
        screen. `ansible-playbook mandala.fleet.state` writes one
        <host>.json snapshot per host into the state dir; we count the
        snapshots freshly written THIS run and report the running tally in
        the top bar, then refresh drift when it finishes. Output is
        captured (writing it to the terminal would corrupt the TUI) and
        surfaced only on failure. Its own worker group so it runs
        alongside the expected-toplevel eval instead of cancelling it."""
        if self._surveying:
            return
        self._surveying = True
        self._survey_n = 0
        self._render_status()  # shows the survey spinner alongside any eval
        directory = drift_mod.state_dir()
        cwd = _ansible_dir()
        env = dict(
            os.environ,
            MANDALA_FLEET_STATE=str(directory),
            PYTHONUNBUFFERED="1",
            ANSIBLE_FORCE_COLOR="0",
        )
        argv = ["ansible-playbook", "mandala.fleet.state"]
        # -1s so a snapshot written in the same clock second as launch
        # still counts as "this run".
        started = time.time() - 1

        def fresh() -> int:
            # Per-host snapshots are <host>.json; skip dotfiles so the
            # .expected.json eval cache (rewritten in this same dir by the
            # concurrent eval worker) is never miscounted as a host.
            if not directory.is_dir():
                return 0
            n = 0
            for p in directory.glob("*.json"):
                if p.name.startswith("."):
                    continue
                try:
                    if p.stat().st_mtime >= started:
                        n += 1
                except OSError:
                    pass
            return n

        def work() -> None:
            out: deque[str] = deque(maxlen=2000)
            try:
                proc = subprocess.Popen(
                    argv,
                    cwd=cwd,
                    env=env,
                    stdin=subprocess.DEVNULL,
                    stdout=subprocess.PIPE,
                    stderr=subprocess.STDOUT,
                    text=True,
                )
            except OSError as e:
                self.call_from_thread(self._survey_done, 0, 1, f"failed to launch: {e}")
                return
            assert proc.stdout is not None
            last = -1
            # Draining stdout (even just to discard) avoids a full-pipe
            # deadlock; each line is also a cheap cue to recount.
            for line in proc.stdout:
                out.append(line.rstrip("\n"))
                n = fresh()
                if n != last:
                    last = n
                    self.call_from_thread(self._survey_progress, n)
            rc = proc.wait()
            err = None if rc == 0 else (out[-1] if out else None)
            self.call_from_thread(self._survey_done, fresh(), rc, err)

        self.run_worker(work, thread=True, exclusive=True, group="survey")

    def _survey_progress(self, n: int) -> None:
        self._survey_n = n
        self._render_status()  # live tally rides the spinner line

    def _survey_done(self, n: int, rc: int, error: str | None = None) -> None:
        self._surveying = False
        self._survey_n = n
        self._fill_drift()
        if rc == 0:
            self._set_status(f"drift refreshed · surveyed {n} host{'' if n == 1 else 's'}")
        else:
            self._set_status(f"survey failed (exit {rc}): {error or ''}".rstrip(), error=True)

    def _drift_done(self, expected: dict[str, str] | None, error: str | None) -> None:
        self._busy = False
        self.expected = expected
        self._fill_drift()
        # _render_status keeps the survey spinner up if it is still counting,
        # so this resting message only surfaces once both jobs are idle; an
        # eval error is sticky and wins over the survey's success message.
        self._set_status(error or "drift refreshed", error=error is not None)

File: crates/mandala-tui/tests/explorer.rs
//! Explorer-tier behavior tests — the `test_tui_mcp.py` analog, driving
//! AppState transitions directly (no terminal, no fleet, no real ansible).
//!
//! Every case here is a design "hard-won behavior": sticky status errors,
//! reload queued-not-dropped, the stale-aggregate guard, the three drift
//! captions, selection→target semantics, the concurrent-jobs spinner line,
//! and the survey's fresh-snapshot tally (dot-files skipped). The survey
//! PIPELINE runs against `sh -c` stubs (the phase-1 runner-test pattern).

use std::collections::BTreeMap;
use std::path::PathBuf;

use chrono::{DateTime, Utc};
use mandala_core::drift::Snapshot;
use mandala_core::inventory::Inventory;
use mandala_tui::event::AppEvent;
use mandala_tui::explorer::{fresh_snapshots, run_survey};
use mandala_tui::state::{AppState, LoadedInventory, Tab, drift_hint};
use serde_json::json;

// ---- fixtures ---------------------------------------------------------------

fn aggregate() -> serde_json::Value {
    json!({
        "schemaVersion": 1,
        "members": {
            "web": {"name": "web", "platform": "metal", "role": "web", "tags": ["edge"]},
            "cache": {"name": "cache", "platform": "metal"},
            "router": {"name": "router", "platform": "opnsense"},
        },
        "groups": {"k3s": ["cache", "web"], "gateway": ["router"]},
        "projections": {"deploy": {"nodes": ["cache", "web"]}},
    })
}

fn loaded(rev: Option<&str>, cached_rev: Option<&str>) -> LoadedInventory {
    let mut cached = BTreeMap::new();
    if cached_rev.is_some() {
        cached.insert(
            "cache".to_string(),
            "/nix/store/cached00-toplevel".to_string(),
        );
    }
    LoadedInventory {
        inventory: Inventory::from_value(aggregate()).expect("fixture aggregate is valid"),
        rev: rev.map(str::to_string),
        cached_rev: cached_rev.map(str::to_string),
        cached,
    }
}

fn now() -> DateTime<Utc> {
    DateTime::parse_from_rfc3339("2026-06-12T12:00:00+00:00")
        .unwrap()
        .with_timezone(&Utc)
}

fn snapshots() -> BTreeMap<String, Snapshot> {
    let snap = |current: &str| -> Snapshot {
        serde_json::from_value(json!({
            "current": current,
            "booted": current,
            "captured_at": "2026-06-12T11:59:00+00:00",
        }))
        .expect("snapshot fixture parses")
    };
    BTreeMap::from([
        ("cache".to_string(), snap("/nix/store/curcache-toplevel")),
        ("web".to_string(), snap("/nix/store/curweb00-toplevel")),
    ])
}

fn filled_state() -> AppState {
    let mut state = AppState::new();
    let req = state.request_load().expect("idle state loads");
    let follow_up = state.on_load_finished(
        req.generation,
        Ok(loaded(Some("aaaaaaaaaaaaaaaa"), None)),
        &snapshots(),
        now(),
    );
    assert!(follow_up.is_none());
    state
}

fn tmp() -> PathBuf {
    use std::sync::atomic::{AtomicU64, Ordering};
    static COUNTER: AtomicU64 = AtomicU64::new(0);
    let dir = std::env::temp_dir().join(format!(
        "mandala-tui-test-{}-{}",
        std::process::id(),
        COUNTER.fetch_add(1, Ordering::Relaxed),
    ));
    std::fs::create_dir_all(&dir).unwrap();
    dir
}

// ---- selection → target (feeds the section-5 actions) -----------------------

#[test]
fn target_is_selection_else_cursor_on_the_active_tab() {
    let mut state = filled_state();
    // No selection: the cursor row (members sort: cache, router, web).
    assert_eq!(state.target().as_deref(), Some("cache"));
    // A multi-selection wins, comma-joined in TABLE order.
    state.members_table.toggle(); // cache
    state.members_table.skip(2); // cursor on web — selection unchanged
    state.members_table.toggle(); // + web
    assert_eq!(state.target().as_deref(), Some("cache,web"));
    // The target follows the ACTIVE tab: groups has its own table/cursor.
    state.tab = Tab::Groups;
    assert_eq!(state.target().as_deref(), Some("gateway"));
    state.groups_table.move_cursor(1);
    assert_eq!(state.target().as_deref(), Some("k3s"));
    // Back on members the selection still holds (selection is per-table).
    state.tab = Tab::Members;
    assert_eq!(state.target().as_deref(), Some("cache,web"));
    // An empty explorer has no target at all.
    assert_eq!(AppState::new().target(), None);
}

// ---- status machinery -------------------------------------------------------

#[test]
fn sticky_error_survives_a_concurrent_success() {
    let mut state = filled_state();
    let (eval, survey) = state.refresh_drift();
    assert!(eval && survey);
    // The eval fails while the survey is still running…
    let follow_up =
        state.on_drift_eval_finished(Err("eval failed: boom".to_string()), &snapshots(), now());
    assert!(follow_up.is_none());
    // …the bar still shows the survey spinner (the error is the RESTING msg).
    assert!(state.status_line().contains("survey"));
    // The survey's success must NOT stomp the sticky error.
    state.on_survey_done(2, 0, None, &snapshots(), now());
    assert_eq!(state.status, "eval failed: boom");
    assert_eq!(state.status_line(), "eval failed: boom");
    // The eval error also dropped the expected set (Python `_drift_done`).
    assert!(state.expected.is_none());
    // The NEXT refresh clears the stickiness…
    let (eval, survey) = state.refresh_drift();
    assert!(eval && survey);
    state.on_survey_done(2, 0, None, &snapshots(), now());
    // …so this survey's success message lands.
    assert_eq!(state.status, "drift refreshed · surveyed 2 hosts");
}

#[test]
fn survey_success_first_keeps_the_eval_spinner_up() {
    let mut state = filled_state();
    let _ = state.refresh_drift();
    // Survey lands first: its resting message is set, but the bar still
    // spins for the eval.
    state.on_survey_done(1, 0, None, &snapshots(), now());
    assert_eq!(state.status, "drift refreshed · surveyed 1 host");
    assert!(state.status_line().starts_with("running   "));
    assert!(state.status_line().contains("eval"));
    // Then the eval error lands and (sticky) replaces the resting message.
    let _ = state.on_drift_eval_finished(Err("eval failed: nope".to_string()), &snapshots(), now());
    assert_eq!(state.status_line(), "eval failed: nope");
}

#[test]
fn spinner_line_lists_every_running_job_with_one_shared_frame() {
    let mut state = filled_state();
    let _ = state.refresh_drift();
    assert_eq!(state.status_line(), "running   ⠋ eval   ·   ⠋ survey");
    state.on_survey_progress(3);
    assert!(state.tick_spinner());
    assert_eq!(
        state.status_line(),
        "running   ⠙ eval   ·   ⠙ survey (3 read)"
    );
    // Idle again: the resting message, and idle ticks change nothing.
    let _ = state.on_drift_eval_finished(
        Ok((Some("aaaaaaaaaaaaaaaa".to_string()), BTreeMap::new())),
        &snapshots(),
        now(),
    );
    state.on_survey_done(3, 0, None, &snapshots(), now());
    assert!(!state.tick_spinner());
    assert_eq!(state.status_line(), "drift refreshed · surveyed 3 hosts");
}

#[test]
fn survey_failure_is_sticky_and_names_the_exit_and_last_line() {
    let mut state = filled_state();
    assert!(state.request_survey());
    state.on_survey_done(0, 2, Some("oops"), &snapshots(), now());
    assert_eq!(state.status, "survey failed (exit 2): oops");
    assert!(state.status_sticky);
    // No captured line: the trailing space is trimmed (Python rstrip).
    let mut state = filled_state();
    assert!(state.request_survey());
    state.on_survey_done(0, 2, None, &snapshots(), now());
    assert_eq!(state.status, "survey failed (exit 2):");
}

#[test]
fn load_failure_surfaces_the_last_error_line_sticky() {
    let mut state = AppState::new();
    let req = state.request_load().unwrap();
    let follow_up = state.on_load_finished(
        req.generation,
        Err("trace line\nerror: attribute missing".to_string()),
        &snapshots(),
        now(),
    );
    assert!(follow_up.is_none());
    assert_eq!(
        state.status,
        "aggregate eval failed: error: attribute missing"
    );
    assert!(state.status_sticky);
    assert!(!state.busy);
}

#[test]
fn eval_expected_never_queues_while_busy() {
    let mut state = filled_state();
    assert!(state.request_eval_expected());
    // A second S while the eval runs is a no-op (returns early on busy)…
    assert!(!state.request_eval_expected());
    // …and does NOT queue anything for later (unlike a reload).
    assert!(!state.reload_pending);
}

// ---- reload queued-not-dropped + the stale-aggregate guard ------------------

#[test]
fn reload_while_busy_queues_and_the_stale_fill_never_paints() {
    let mut state = AppState::new();
    let first = state.request_load().expect("initial load starts");
    assert_eq!(first.generation, 0);
    // `r` while the eval worker runs: queued, not dropped; the inventory is
    // rebound (unevaluated → no deploy nodes) and expected dropped.
    assert!(state.request_reload().is_none());
    assert!(state.reload_pending);
    assert_eq!(state.generation, 1);
    assert!(state.inventory.is_none());
    assert!(state.deploy_nodes().is_empty());
    // The FIRST load lands, carrying the superseded generation: it must not
    // paint, and the queued reload starts with the fresh generation.
    let follow_up = state.on_load_finished(
        first.generation,
        Ok(loaded(Some("aaaaaaaaaaaaaaaa"), None)),
        &snapshots(),
        now(),
    );
    assert_eq!(follow_up.expect("queued reload starts").generation, 1);
    assert!(state.member_rows.is_empty()); // the stale aggregate never painted
    assert!(state.busy); // the queued reload is now in flight
    assert!(!state.reload_pending);
    // The fresh load paints normally.
    let follow_up = state.on_load_finished(
        1,
        Ok(loaded(Some("bbbbbbbbbbbbbbbb"), None)),
        &snapshots(),
        now(),
    );
    assert!(follow_up.is_none());
    assert_eq!(state.member_rows.len(), 3);
    assert_eq!(
        state.status,
        "3 members, 2 groups — space/shift+↑↓ select · p ping · R reboot · D deploy"
    );
}

#[test]
fn reload_queued_behind_an_expected_eval_runs_after_it() {
    let mut state = filled_state();
    assert!(state.request_eval_expected());
    // `r` during the expected eval queues (busy covers BOTH eval classes).
    assert!(state.request_reload().is_none());
    assert!(state.reload_pending);
    // The eval settles → the queued reload starts.
    let follow_up = state.on_drift_eval_finished(
        Ok((Some("aaaaaaaaaaaaaaaa".to_string()), BTreeMap::new())),
        &snapshots(),
        now(),
    );
    assert_eq!(follow_up.expect("queued reload starts").generation, 1);
}

// ---- drift captions (the exact three cases) ---------------------------------

#[test]
fn drift_caption_expected_fresh() {
    let mut state = filled_state();
    assert!(state.request_eval_expected());
    let _ = state.on_drift_eval_finished(
        Ok((
            Some("aaaaaaaaaaaaaaaa".to_string()),
            BTreeMap::from([(
                "cache".to_string(),
                "/nix/store/curcache-toplevel".to_string(),
            )]),
        )),
        &snapshots(),
        now(),
    );
    assert_eq!(
        state.drift_hint,
        "S refresh drift (survey + eval) · R reboot a reboot-pending row   expected @ aaaaaaaaaaa"
    );
    // The rev-keyed cache also satisfies this case on a plain load: same
    // clean rev → the cached expectation is adopted.
    let mut state = AppState::new();
    let _ = state.request_load();
    let _ = state.on_load_finished(
        0,
        Ok(loaded(Some("cccccccccccccccc"), Some("cccccccccccccccc"))),
        &snapshots(),
        now(),
    );
    assert!(state.expected.is_some());
    assert!(state.drift_hint.ends_with("expected @ ccccccccccc"));
}

#[test]
fn drift_caption_contract_moved_since_last_eval() {
    let mut state = AppState::new();
    let _ = state.request_load();
    let _ = state.on_load_finished(
        0,
        Ok(loaded(Some("bbbbbbbbbbbbbbbb"), Some("aaaaaaaaaaaaaaaa"))),
        &snapshots(),
        now(),
    );
    assert!(state.expected.is_none()); // stale cache is NOT adopted
    assert_eq!(
        state.drift_hint,
        "S refresh drift (survey + eval) · R reboot a reboot-pending row   \
         contract MOVED since last eval (cache @ aaaaaaaaaaa, repo @ bbbbbbbbbbb) — press S"
    );
}

#[test]
fn drift_caption_never_evaluated() {
    let state = filled_state();
    assert_eq!(
        state.drift_hint,
        "S refresh drift (survey + eval) · R reboot a reboot-pending row   \
         (expected NOT evaluated yet — press S)"
    );
    // The pure caption fn covers the dirty-rev spelling too.
    assert_eq!(
        drift_hint(true, Some(&format!("{}-dirty", "a".repeat(40))), None),
        "S refresh drift (survey + eval) · R reboot a reboot-pending row   \
         expected @ aaaaaaaaaaa-dirty"
    );
}

// ---- the survey tally + pipeline (sh -c stubs, never real ansible) ----------

#[test]
fn survey_tally_counts_only_fresh_non_dot_snapshots() {
    let dir = tmp();
    std::fs::write(dir.join("alpha.json"), "{}").unwrap();
    std::fs::write(dir.join("beta.json"), "{}").unwrap();
    std::fs::write(dir.join(".expected.json"), "{}").unwrap(); // the eval cache
    std::fs::write(dir.join("notes.txt"), "x").unwrap();
    let past = std::time::SystemTime::UNIX_EPOCH;
    assert_eq!(fresh_snapshots(&dir, past), 2);
    // Snapshots older than the run start don't count as "this run".
    let future = std::time::SystemTime::now() + std::time::Duration::from_secs(3600);
    assert_eq!(fresh_snapshots(&dir, future), 0);
    // A missing dir counts zero (the Python `is_dir` guard).
    assert_eq!(fresh_snapshots(&dir.join("nope"), past), 0);
}

async fn collect_survey(argv: Vec<String>, state_dir: PathBuf) -> Vec<AppEvent> {
    let (tx, mut rx) = tokio::sync::mpsc::channel(64);
    let cwd = std::env::temp_dir();
    tokio::spawn(run_survey(tx, argv, cwd, state_dir));
    let mut events = Vec::new();
    loop {
        let ev = tokio::time::timeout(std::time::Duration::from_secs(10), rx.recv())
            .await
            .expect("survey settles in time")
            .expect("survey channel stays open until done");
        let done = matches!(ev, AppEvent::SurveyDone { .. });
        events.push(ev);
        if done {
            break;
        }
    }
    events
}

#[tokio::test]
async fn survey_pipeline_tallies_fresh_snapshots_and_settles_clean() {
    let state_dir = tmp();
    let script = r#"echo starting; printf '{}' > "$MANDALA_FLEET_STATE/hostx.json"; echo done"#;
    let events = collect_survey(
        vec!["sh".to_string(), "-c".to_string(), script.to_string()],
        state_dir,
    )
    .await;
    let Some(AppEvent::SurveyDone { n, rc, error }) = events.last() else {
        panic!("survey did not settle: {events:?}");
    };
    assert_eq!((*n, *rc), (1, 0));
    assert!(
        error.is_none(),
        "clean survey surfaces no output: {error:?}"
    );
    // The live tally reported the freshly written snapshot before settling.
    assert!(
        events
            .iter()
            .any(|e| matches!(e, AppEvent::SurveyProgress { n: 1 })),
        "no progress tally in {events:?}"
    );
}

#[tokio::test]
async fn survey_pipeline_surfaces_the_last_line_only_on_failure() {
    let events = collect_survey(
        vec![
            "sh".to_string(),
            "-c".to_string(),
            "echo oops >&2; exit 3".to_string(),
        ],
        tmp(),
    )
    .await;
    let Some(AppEvent::SurveyDone { n, rc, error }) = events.last() else {
        panic!("survey did not settle: {events:?}");
    };
    assert_eq!((*n, *rc), (0, 3));
    assert_eq!(error.as_deref(), Some("oops"));
}

#[tokio::test]
async fn survey_pipeline_reports_a_launch_failure() {
    let events = collect_survey(vec!["/nonexistent/mandala-survey-stub".to_string()], tmp()).await;
    let Some(AppEvent::SurveyDone { n, rc, error }) = events.last() else {
        panic!("survey did not settle: {events:?}");
    };
    assert_eq!((*n, *rc), (0, 1));
    assert!(
        error
            .as_deref()
            .is_some_and(|e| e.starts_with("failed to launch: ")),
        "unexpected error: {error:?}"
    );
}

// ---- unevaluated inventory never evals on the UI path -----------------------

#[test]
fn deploy_nodes_are_empty_until_the_background_eval_lands() {
    let mut state = filled_state();
    assert_eq!(state.deploy_nodes(), ["cache", "web"]);
    // `r` rebinds an unevaluated inventory: no nodes, and a drift repaint
    // (e.g. a landing survey) renders an EMPTY table rather than forcing an
    // eval on the UI thread.
    let _ = state.request_reload();
    assert!(state.deploy_nodes().is_empty());
    state.fill_drift(&snapshots(), now());
    assert!(state.drift_rows.is_empty());
}
